### DISCUSSION TITLE
Not enough cache hit

### DISCUSSION THREAD

**Discussion by BlzsDnl** [ccache/ccache#175] [Support]
Dear Ccache Developers!

We have an issue in one of our projects with ccache, and i would like to ask for your opinion about it.

We have two linux projects and ccache works perfectly on the first one, but not on the other one. We compile them on the same machine, with same gcc version but with different flags. The sources also have common parts (whole files).

In the case of the problematic project, there aren’t enough cache hit, so the speed of compiling is about 95% for the second compiling (The first compiling is about 110%). Statistics:

After the first run:
```
cache hit (direct)                     1
cache hit (preprocessed)               0
cache miss                          3642
cache hit rate                      0.03 %
called for link                      136
called for preprocessing            1809
no input file                          1
cleanups performed                    11
files in cache                      5878
cache size                           7.3 GB
max cache size                      10.0 GB
```

After the second run:
```
cache hit (direct)                   321
cache hit (preprocessed)             239
cache miss                          6322
cache hit rate                      8.14 %
called for link                      272
called for preprocessing            3385
no input file                          2
cleanups performed                    55
files in cache                      5289
cache size                           7.6 GB
max cache size                      10.0 GB
```

The log file is full with „no such manifest file” and „result: cache miss” entries, i attached a part of it to this mail.
I have tried a lot of config combination, but non of it helped. The current config is in the log file.
[log](https://github.com/ccache/ccache/files/985881/ccache2.txt)

Can you suggest something about this?

**Comment by afbjorklund** [comment]
It doesn't sound like you're hitting any of the "not supported" cases, so looks like regular old cache misses.
There aren't too many debugging tools available through config, so it's hard to tell why the hashes differ.

As far as I could tell, your ccache configuration was default too. (except paths and sizes, and such)
So nothing much that can be done, you'll need to compare command lines and preprocessor output ?

i.e. the `Executing <private_information>`

Sometimes people add timestamps or similar entropy, to global macros and global headers and such ?
That will cause cache misses. If it's just added in a comment or something, it should still hit after cpp.

I gather it's not just the `base_dir` ($PWD) ?

**Comment by afbjorklund** [comment]
@BlzsDnl:
Note that `hash_dir` now defaults to true, so if you are generating debuginfo (`-g`) from a temporary working directory (mktemp -d) you _will_ get random strings in your binaries - and lots of cache misses.

i.e. the `Working directory: <private_information>`

This happened to the good people from Spack in #173, and they used the `hash_dir = false` workaround until they have relocated their debuginfo. This can be done with the `-fdebug-prefix-map=old=new` arg.

**Comment by BlzsDnl** [comment]
It looks like you are right and the debuginfo generating causes the problem, because without the `-g`  flag the build time is only 20% for the second run.

However, the `hash_dir = false` doesn't help. I will try the `-fdebug-prefix-map=old=new` and run some tests.
Please do not close this issue until that.

**Comment by afbjorklund** [comment]
For best effect, also set `base_dir`

**Comment by BlzsDnl** [comment]
My information was misleading, sorry. The problem wasn't the -g flag, or any of the flags, the faster build was only consequence.

The real problem will be the cache size, because for the second run ccache cleaned the hashed files. Background information: we build with jenkins and set the cache dir of ccache to an other one, depending on the project. Directory tree: .../ccache_cache/`project specific dir`/cached hashs and dirs

Original test:
```
Max cache size config parameter: 10G
Project size in statistics: 7.6G
Project size for real: 7.1G
Size of all projects for real: 9.6G
Speed of project: 95%
```

Second test:
```
Max cache size config parameter: 30G
Project size in statistics: 12.0G
Size of all projects for real: 20G
Speed of project: 45%
```

Third test:
```
Max cache size config parameter: 40G
Project size in statistics: 14.4G
Project size for real: 14G
Size of all projects for real: 22G
Speed of project: 45%
```

That's why the `hash_dir = false` didn't help, because the problem wasn't the debuginfo generating. However, the difference between the 20% (without -g) and 45% (with -g) may really because of debuginfo. I will run other tests again and write to you, but something is definitely wrong with cache size.

**Comment by afbjorklund** [comment]
Actually, that is why the `cleanups performed` metric was added - so that you could see the number of "garbage collections" being done, to know if you need to increase the cache size...
You could also try running with `compression = true`, since that helps a lot with the debuginfo. It will increase the effective size of your cache. Typically something like 150% or so ?
